### PR TITLE
 [WIP] [2/4] Intel GPU Runtime Upstreaming for Device

### PR DESCRIPTION
**Pull request by guangyey**
# Motivation
According to [[1/4] Intel GPU Runtime Upstreaming for Device](https://github.com/pytorch/pytorch/pull/116019), the second PR  covers the changes under `aten`.

# Design
We will compile the code for XPU separately into a library named `libtorch_xpu.so`. Currently, it primarily offers device-related APIs, including
- `getCurrentDeviceProperties`
- `getDeviceProperties`
- `getGlobalIdFromDevice`
- `getDeviceFromPtr`

# Additional Context
`XPUHooks` is an indispensable part of the runtime. We upstream `XPUHooks` in this PR since there is some code related to `Device` in it and we also refine some logic and code to avoid forward declaration in `DLPack`.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Comment by guangyey** [comment]
@gujinghui @jgong5 @EikanWang @arthuryuan1987 please review.

File: aten/src/ATen/DLConvertor.cpp [hunk 1]
```diff
@@ -101,8 +101,9 @@ static DLDevice getDLDevice(const Tensor& tensor, const int64_t& device_id) {
       ctx.device_type = DLDeviceType::kDLROCM;
       break;
     case DeviceType::XPU:
-      ctx = at::detail::getXPUHooks().getDLPackDeviceFromATenDevice(
-          ctx, tensor.device(), tensor.data_ptr());
+      ctx.device_type = DLDeviceType::kDLOneAPI;
+      ctx.device_id =
+          at::detail::getXPUHooks().getGlobalIdFromDevice(tensor.device());
```
Comment (jgong5) [inline aten/src/ATen/DLConvertor.cpp:106]: Why do we have to override `ctx.device_id` here?

File: aten/src/ATen/xpu/XPUContext.cpp [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+DeviceProp* getCurrentDeviceProperties() {
+  auto device = c10::xpu::current_device();
+  return getDeviceProperties(device);
+}
+
+DeviceProp* getDeviceProperties(int device) {
+  c10::call_once(init_flag, initXPUContextVectors);
+  if (device == -1)
+    device = c10::xpu::current_device();
+  TORCH_CHECK(
```
Comment (jgong5) [inline aten/src/ATen/xpu/XPUContext.cpp:66]:
```suggestion
    TORCH_CHECK(
```

File: aten/src/ATen/xpu/detail/XPUHooks.cpp [hunk 3]
```diff
@@ -0,0 +1,35 @@
+#include <ATen/xpu/XPUContext.h>
+#include <ATen/xpu/XPUDevice.h>
+#include <ATen/xpu/detail/XPUHooks.h>
+#include <c10/util/CallOnce.h>
+
+namespace at::xpu::detail {
+
+void XPUHooks::initXPU() const {
+  // TODO:
```
Comment (EikanWang) [inline aten/src/ATen/xpu/detail/XPUHooks.cpp:9]: Add comprehensive TODO comments here.

File: aten/src/ATen/DLConvertor.cpp [hunk 4]
```diff
@@ -129,7 +130,7 @@ static Device getATenDevice(const DLDevice& ctx, void* data) {
       return at::Device(DeviceType::HIP, ctx.device_id);
 #endif
     case DLDeviceType::kDLOneAPI:
-      return at::detail::getXPUHooks().getATenDeviceFromDLPackDevice(ctx, data);
+      return at::detail::getXPUHooks().getDeviceFromPtr(data);
```
Comment (EikanWang) [inline aten/src/ATen/DLConvertor.cpp:133]: May I know why not `at::Device(DeviceType::XPU, ctx.device_id)`?

File: aten/src/ATen/DLConvertor.cpp (hunk 1 above)
Comment (guangyey) [reply aten/src/ATen/DLConvertor.cpp:106]: for dlpack, we use the device index enumerated from sycl::device::get_devices(), aka global id, as the parameter. So we need to convert the device index of the framework into a global id here.

File: aten/src/ATen/DLConvertor.cpp (hunk 4 above)
Comment (guangyey) [reply aten/src/ATen/DLConvertor.cpp:133]: at::Device is constructed by the device index of framework presenting the order index of GPU device. But `ctx.device_id` is a global order index containing GPU device or other device.

**Review by EikanWang — APPROVED** [review]

**Comment by guangyey** [comment]
@ezyang please help review.

File: aten/src/ATen/xpu/XPUContext.cpp (hunk 2 above)
Comment (guangyey) [reply aten/src/ATen/xpu/XPUContext.cpp:66]: according to lintuner's rule.

File: aten/src/ATen/xpu/detail/XPUHooks.cpp (hunk 3 above)
Comment (guangyey) [reply aten/src/ATen/xpu/detail/XPUHooks.cpp]: done.

**Review by jgong5 — APPROVED** [review]

**Comment by guangyey** [comment]
close this PR, and resubmit PR in [#116833](https://github.com/pytorch/pytorch/pull/116833) via ghstack.

**Pull request closed without merging**